src/rt/minfo.d

```diff
@@ -60,6 +60,339 @@ struct ModuleGroup
+                    break;
+distloop:
+                // search for next (previous) module in cycle.
+                foreach(int m, d; distance)
+                {
+                    if(d == curdist)
+                    {
+                        // determine if m can reach target
+                        foreach(e; edges[m])
+                        {
+                            if(e == target)
+                            {
+                                // recurse
+                                target = m;
+                                break distloop;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+
+        // this function initializes the bookeeping necessary to create the
+        // cycle path, and then creates it. It is a precondition that src and
+        // target modules are involved in a cycle
+        void genPath(int srcidx, int targetidx)
+        {
+            assert(srcidx != -1);
+            assert(targetidx != -1);
+
+            // set up all the arrays. Use the GC, we are going to exit anyway.
+            distance.length = len;
+            edges.length = len;
+            foreach(i, m; _modules)
+            {
+                // use reachable, because an import can appear more than once.
+                // https://issues.dlang.org/show_bug.cgi?id=16208
+                reachable[] = 0;
+                foreach(e; m.importedModules)
+                {
+                    auto impidx = findModule(e);
+                    if(impidx != -1 && impidx != i)
+                        reachable[impidx] = 1;
+                }
+
+                foreach(int j, r; reachable)
+                {
+                    if(r)
+                        edges[i] ~= j;
+                }
+            }
+
+            // a cycle starts with the source.
+            cyclePath ~= srcidx;
+            // first get to the target
+            shortest(srcidx, targetidx);
+            // now get back.
+            shortest(targetidx, srcidx);
+        }
+
+        // find all the non-trivial dependencies (that is, dependencies that have a
+        // ctor or dtor) of a given module.  Doing this, we can 'skip over' the
+        // trivial modules to get at the non-trivial ones.
+        size_t _findDependencies(int idx, bool orig = true)
+        {
+            if(reachable[idx])
+                return 0;
+            auto current = _modules[idx];
+            size_t result = 0;
+            reachable[idx] = 1;
+            if(!orig && (flags[idx] & (MIctor | MIdtor)) && !(flags[idx] & MIstandalone))
+                // non-trivial, stop here
+                return result + 1;
+            foreach (m; current.importedModules)
+            {
+                auto midx = findModule(m);
+                if(midx != -1)
+                    // not part of this DSO, don't consider it.
+                    result += _findDependencies(midx, false);
```

Comment:
> Should check reachable before recursing.

yes, I can do this. And you are right, I can switch from flags array to bit arrays like you had before.

> Might make sense to replace the recursion with a loop and an allocated stack to save on stack usage.

I can do this. It will save runtime to avoid calls as well. Largest stack we need in this case is the size of the module array.